### ISSUE TITLE
Installing and loading shared dependencies with different newest matching versions fails

### ISSUE THREAD

**Issue by mpeterv** [luarocks/luarocks#647]
Let's say that rock `A` depends on rock `B` several times indirectly through other rocks, possible with a direct dependency, too. If some of these dependencies limit max version (e.g. `B >= 0.1.0, < 0.2.0`) some things fail when installing `A` and when loading `A` and its dependencies with `luarocks.loader`.

When installing `A`, all dependencies on `B` will be somehow satisfied, possibly by different versions. However, satisfying each dependency may remove previously installed version of `B`. This is inefficient and may result in missing dependencies if there is no single version of `B` that satisfies all dependencies. E.g. if `A` depends on `B >= 0.1.0, < 0.2.0` and `C`, where `C` depends on `B >= 0.2.0, < 0.3.0`, then installing `A` will first install `B 0.1.0`, then install `B 0.2.0` and remove `B 0.1.0`, bypassing broken dependency check as `A` is not installed by that time.

When loading `A` with `luarocks.loader`, if dependencies on `B` are satisfied with different versions, then `add_context` may select a version of `B` that does not satisfy all the dependencies.

I think that luarocks should maintain that for each installed package, if it depends on another package with several constraints coming from intermediate packages, there should be at least one installed version satisfying all of these constraints. luarocks.loader should use this when selecting versions. Currently luarocks maintains that each constraint is satisfied by some version, maybe using different ones for each constraint, which happens to be same as the stronger invariant if all dependencies are in open ended `B >= 0.x.0` format.

**Comment by hishamhm** [comment]
> E.g. if A depends on B >= 0.1.0, < 0.2.0 and C, where C depends on B >= 0.2.0, < 0.3.0, then installing A will first install B 0.1.0, then install B 0.2.0 and remove B 0.1.0, bypassing broken dependency check as A is not installed by that time.

Since the dependency chain of this example is broken by design (no single version of `B` satisfies all dependencies) there's nothing we can do to get a correct result anyway. Keeping both B 0.1.0 and 0.2.0 around does not solve the problem.

If the set of dependency constraints is no longer disjoint, the problem will no longer happen since the intersection will be naturally picked.

E.g. if `A` depends on `B >= 0.1.0, <= 0.2.0` and `C`, where `C` depends on `B >= 0.2.0, < 0.3.0`, then this means `B 0.2.0` already exists, so installing `A` will first install `B 0.2.0` and everything will work. 

However, if `C` bumps up the dependency requirements for `B` in a later upgrade, the system will note that it had to keep `B 0.2.0` around for `A`, and that would be a red-flag that the dependency chain is now broken.

The red-flag scenario, as far as I can tell, is then: upgrading C (possibly indirectly) bumped the version of B but required to keep an older version of B around for D's sake, and both C and D happen to be (possibly indirect) dependencies of A (and D may be equal to A).

_However_, the real problem here is that `A` was overly optimistic by not specifying a safe version range for `C`. If A did specify, upgrading `C` would keep the old safe version of `C` around, with the old safe version for `B`. These would be picked up by luarocks.loader and everything would work.

If we were to change something related to this, it would be to provide a stronger warning when that red-flag scenario happens at _install_ time. And the best thing to do would be to present a message like `"dependency specifications for "..A.." are inconsistent after installing "..C.." "..C.rockspec.version.." — please contact the maintainer for "..A.." at "..A.rockspec.maintainer`

But I wouldn't complicate the behavior of the loader to cater for the broken case. By the time the loader runs, the installation should have things sorted appropriately.

Is this a theoretical concern or have you seen this in the wild? 

Still, thinking about it, I don't think the loader does the right thing if the dependency for `A` goes `B; C` with `B` dependening on `X >= 1, <= 3` and `C` depending on `X >=2, <= 4`. I _think_ it will end up working or not depending on the order that the `require`s appear inside `A`. I understand that this is what you were aiming at in your last paragraph; but I consider this and the "broken" case to be separate issues.

**Comment by mpeterv** [comment]
> However, the real problem here is that A was overly optimistic by not specifying a safe version range for C. If A did specify, upgrading C would keep the old safe version of C around, with the old safe version for B.

Yes, I think that using precise ranges for all dependency constraints works very well. As long as rockspec author actually tests his rock it should be foolproof. Almost no one uses this style though.

> Is this a theoretical concern or have you seen this in the wild? 

It is theoretical but there is at least one popular rock (busted) that uses a range (a fixed version even) as a constraint. I think that eventually promoting version ranges as constraints based would be nice. I would feel much better about making a new major release of some theoretical library with breaking changes if I knew that most rocks depending on it wouldn't try to use the new version and break. But it's certainly nothing urgent, just something I feel should be an the radar for things to change in LR3, potentially.

> I understand that this is what you were aiming at in your last paragraph; but I consider this and the "broken" case to be separate issues.

Yes, I've opened this to discuss a little bit first before deciding what to do. So, the points are:

* Installing a rock should check that all dependencies are "consistent" (satisfied by a single installed version), if they are not, a warning is shown after installation (disjoint constraints case).
* Luarocks loader should pick version that matches all constraints, regardless of `require` order (do not check for disjoint constraints case).

There is still a minor issue of installing several versions, replacing each other, even in the "good" case.
E.g. in your last example with rocks `A`, `B`, `C`, and `X`, if `A` dependencies are in the other order (`C; B`) then `X == 4` will be installed first for `C` and then will be replaced by `X == 3` for `B`. The final state is correct but it's a bit inefficient.

**Comment by hishamhm** [comment]
Agreed on both main points.

My gut feeling is that if those are solved, that third minor issue will be solved for free (because checking for consistency implies traversing and fixing the loader implies sorting).
